### PR TITLE
Add delete binding (unbind) action to provisioned services

### PR DESCRIPTION
**Pull request by benjaminapetersen**
Depends on #1578

WIP, still some things to do.  Based on [mock-ups from here](https://openshift.github.io/openshift-origin-design/web-console/3-project-details/binding-in-project).  Progress thus far:

screen shot 2017-05-03 at 12 10 26 pm

screen shot 2017-05-03 at 12 10 34 pm

screen shot 2017-05-03 at 12 10 39 pm

TODO:
- [ ] `<service>` has been unbound from **`<application>`**.  Application name is missing.
- [x] factor `bindService` and `unbindService` duplicate code into separate service, OR collapse `unbindService` into bindService & toggle.

@jwforres @spadgett @ncameronbritt

### CODE REVIEW

File: app/scripts/directives/bindService.js [hunk 1]
```diff
@@ -16,10 +17,16 @@ angular.module('openshiftConsole').component('bindService', {
 });
 
 function BindService($filter,
+                     BindingModalUtils,
                      DataService,
                      DNS1123_SUBDOMAIN_VALIDATION) {
   var ctrl = this;
```
Comment (benjaminapetersen) [inline app/scripts/directives/bindService.js]: @jwforres thoughts here, pulled some things from `bindService` to share w/both components.

File: app/scripts/services/bindingModalUtils.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+angular.module('openshiftConsole')
+  .factory('BindingModalUtils', function() {
```
Comment (spadgett) [inline app/scripts/services/bindingModalUtils.js]: @benjaminapetersen We're planning to switch to the angular-patternfly pf-wizard component, which should handle most of this for us.

Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: Sounds good, keep the service till then so don't have to duplicate code?

Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: Going to probably push up `sortApplications` as well, as I need that for `unbindService` to be able to get the name of the associated app.

**Comment by spadgett** [comment]
I feel like this should be either create binding / delete binding OR add binding / remove binding. Create / remove seems wrong.

@ncameronbritt @jwforres Opinion?

**Comment by benjaminapetersen** [comment]
agree... anyone have a preference?

**Comment by ncameronbritt** [comment]
I agree with @spadgett about choosing the right pair of verbs. I think I like create/delete better as well, but I also want to make sure that we're choosing words that reflect how we want users to understand what they're doing. Ignoring what happens technically, is there a metaphor we would use to describe the concept?

**Comment by benjaminapetersen** [comment]
screen shot 2017-05-09 at 11 40 39 am

**Comment by benjaminapetersen** [comment]
btw, rebased

**Comment by benjaminapetersen** [comment]
[TEST]

**Comment by benjaminapetersen** [comment]
ah, meh, another rebase todo.

**Comment by benjaminapetersen** [comment]
@jwforres rebased 

ServiceInstancesService... ftw?

**Comment by benjaminapetersen** [comment]
Will rebase again after #1599

File: app/views/directives/bind-service/remove-binding-confirm.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      <strong>{{ctrl.selectedBinding}}</strong> has been unbound
+      <!-- TODO: will need to share the sortApplications() function from bindService (and the resource requests block) -->
+      <!-- from <strong>{{ctrl.target.metadata.name}}</strong>-->
+    </h3>
+    <p class="mar-top-lg">
+      <span class="pficon pficon-info"></span> You will need to redeploy your pods for the un-binding to take effect.
+    </p>
+  </div>
+  <div ng-if="ctrl.error">
+    <div class="title">Removal of Binding Failed <span class="fa fa-times text-danger"></span></div>
```
Comment (jwforres) [inline app/views/directives/bind-service/remove-binding-confirm.html]: if we called the action Delete Binding, then it shouldnt say Removal here

File: app/views/directives/bind-service/remove-binding-confirm.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    <div class="sub-title">
+      <span ng-if="ctrl.error.data.message">
+        {{ctrl.error.data.message | upperFirst}}
+      </span>
+      <span ng-if="!ctrl.error.data.message">
+        An error occurred removing the binding.
```
Comment (jwforres) [inline app/views/directives/bind-service/remove-binding-confirm.html]: same here, delete not remove

File: app/views/directives/bind-service/select-binding.html [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<h3 class="mar-top-none">
+  Select a service to unbind from <strong>{{ctrl.displayName}}</strong>
```
Comment (jwforres) [inline app/views/directives/bind-service/select-binding.html]: we agreed on friday we didnt want to use the term "unbind"

File: app/scripts/services/bindingModalUtils.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                                    statefulSets) {
+          // Don't waste time sorting on each data load, just sort when we have them all
+          if (deploymentConfigs && deployments && replicationControllers && replicaSets && statefulSets) {
+            var apiObjects = deploymentConfigs.concat(deployments)
+                                              .concat(replicationControllers)
+                                              .concat(replicaSets)
+                                              .concat(statefulSets);
+            return _.sortByAll(apiObjects, ['metadata.name', 'kind']);
+          }
+        }
```
Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: I don't think `_.sortByAll` will ever return a null.  Even `_.sortByAll(null, ['metadata.name', 'kind'])` will return `[ ]`.

**Comment by jwforres** [comment]
no i mean the fact that you have an if check here, in the case where the if
doesn't pass the function returns nothing, effectively an undefined result
wherever it was called from

File: app/scripts/services/bindingModalUtils.js (hunk 6 above)
Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: Ahh... my bad. Got it.  Will do.

**Comment by benjaminapetersen** [comment]
Fixed by PR #1705

**Pull request closed without merging**